Overhaul certain spack packages

johnwparent opened the pull request:
This PR ports adds Windows support to a number of packages and ports a number of packages to use the recent builder system merged in by #30738 

This PR also adds two new builders `NMakefiles` and `MSBuild`

You may also note that all of these packages are dependencies of Paraview/VTK. This is not a coincidence as this PR is part of a larger effort to support those two packages on Windows.

This PR ports:
Python - to multibuild system
Perl - to multibuild system
adios2 -  to Windows and multibuild system
c-blosc -  to Windows and multibuild system
double-conversion -  to  multibuild system
eigen -  to  multibuild system
expat - to Windows and multibuild system
freetype - to multibuild system
jsoncpp - to multibuild system
libjpeg-turbo - to multibuild system
libogg -  to Windows and multibuild system
libpng -  to multibuild system
libtheora -  to Windows and multibuild system
libxml2 - to Windows and multibuild system
lz4 -  to Windows and multibuild system
netcdf-cxx - to multibuild system
netcdf-c -  to Windows and multibuild system
pegtl -  to Windows and multibuild system
proj -  to Windows and multibuild system
protobuf -  to Windows and multibuild system
snappy -  to Windows and multibuild system
winlibiconv - created a new package to provide libiconv on Windows
xz -  to Windows and multibuild system
zfp -  to Windows and multibuild system
zstd -  to Windows and multibuild system


This pr depends on #31141 and #33021

johnwparent committed:
port libtiff

update libtiff

johnwparent committed:
Port libpng

libpng - multi build system

johnwparent committed:
port expat

expat style

johnwparent committed:
lz4

lz4 serial on Windows

lz4 - multi buildsystem

johnwparent committed:
libogg port

libogg - multi build system

johnwparent committed:
jpeg pass

turbo

johnwparent committed:
protobuf

protobuf - multi build system

Fixup protobuf

style protobuf

update protobuf

johnwparent committed:
Add winiconv

Split iconv phases

winlibiconv - multi build system

Update libiconv and winlibiconv

winlibiconv generic builder

adamjstewart commented on var/spack/repos/builtin/packages/libtiff/package.py:
```diff
@@ -6,7 +6,7 @@
 
 
-class Libtiff(AutotoolsPackage):
+class Libtiff(CMakePackage):
```
This should be changed to use Spack's new multi-build system support. See https://spack.readthedocs.io/en/latest/packaging_guide.html#multiple-build-systems for details.

johnwparent replied:
@adamjstewart I was unaware of this until about five minute ago. I've added a note to the top of this PR. I'll be refactoring as much of this PR as possible to use the multiple build systems. Feel free to hold of on review until that point, or review the in-situ version, whatever works.

adamjstewart replied:
Yep, this was merged just yesterday. Should make Windows support significantly easier since packages can subclass AutotoolsPackage only on Linux/macOS and use a different build system on Windows. Hopefully we can refactor most packages like this.

johnwparent replied:
Personally I'm very excited about this PR. As you can see a number of these ports ended up fairly messy. I'll be refactoring as much as possible as I go.

adamjstewart replied:
Same, packages like Python can also be refactored as well.

johnwparent replied:
I've refactored Python and Perl  and included them here.

adamjstewart reviewed (changes requested):
It would be helpful if each of these packages were changed in separate PRs since each change doesn't require other changes.

johnwparent committed:
Refactor adios2 to new builder

Finalize adios2

adios2 style

update adios2

johnwparent committed:
perl - multi build system

Update perls new builders

style perl

Perl: Factor out common setup logic

Factor common setup logic accross Windows/Posix to be in one class
Perl invalid ref fix

perl: correct nmake import

johnwparent committed:
python - multi build system

update python

update python

Finish python package scoping

Python style

johnwparent commented:
@alalazo I've used your PR to port/refactor a number of packages. If you have a second I'd appreciate a review just to be sure I'm using this new functionality as optimally as possible.

adamjstewart commented:
Can you split this PR into separate PRs for each package? Most of these changes are independent of each other, and it's really hard to properly review PRs that modify dozens of packages.

alalazo commented on lib/spack/spack/build_systems/nmakefile.py:
```diff
@@ -0,0 +1,126 @@
+)
+
+
+class NMakePackage(spack.package_base.PackageBase):
+    """Specialized class for packages built using a Makefiles."""
+
+    #: This attribute is used in UI queries that need to know the build
+    #: system base class
+    build_system_class = "NMakePackage"
+
+    build_system("nmake")
+    conflicts("platform=linux",  when="build_system=nmake")
+    conflicts("platform=darwin", when="build_system=nmake")
+    conflicts("platform=cray",   when="build_system=nmake")
```
There's already a `nmake.py` file. Is this different? In case can you extend the other one?

alalazo commented on lib/spack/spack/build_environment.py:
```diff
@@ -558,6 +558,9 @@ def _set_variables_for_single_module(pkg, module):
 
     if sys.platform == "win32":
         m.nmake = Executable("nmake")
+        m.msbuild = Executable("msbuild")
+        # analog to configure for win32
+        m.cscript = Executable("cscript")
```
We'll be trying to remove these global settings here, in favor of using `setup_dependent_package` in the corresponding builder. In this way we won't pick-up executables from the system accidentally (or at least it will be rarer).

alalazo reviewed (changes requested):
I have a comment on `nmake` and on adding more global executables. Also, can we start extracting single package changes from here? I have seen packages, for instance, that don't need modifications to core and can be reviewed on their own.

scheibelp commented on lib/spack/spack/build_environment.py:
```diff
@@ -287,7 +287,7 @@ def clean_environment():
     keep_flags = set()
     # set of pairs
-    replace_flags = []  # type: List[Tuple[str,str]]
+    replace_flags = []
```
Was this type information incorrect?

johnwparent replied:
No, it was causing a CI precheck failure.

adamjstewart replied:
@tgamblin fixed this in a separate PR

scheibelp commented on var/spack/repos/builtin/packages/freetype/package.py:
```diff
@@ -3,10 +3,12 @@
 
 
-class Freetype(AutotoolsPackage):
+class Freetype(CMakePackage):
```
Either:

* This should inherit both CMakePackage and AutotoolsPackage and define the build_system variant
* or (if CMake is now used for freetype on both Windows and Linux) the Autotools references should be removed

johnwparent replied:
Autotools has been removed.

scheibelp commented on var/spack/repos/builtin/packages/libogg/package.py:
```diff
@@ -3,10 +3,16 @@
 
 
-class Libogg(AutotoolsPackage):
+class Libogg(Package):
```
This should inherit CMakePackage, Package, and AutotoolsPackage (whenever it's possible for a package to need a build system, it should be included here).

johnwparent replied:
Understood, amended.

scheibelp commented on var/spack/repos/builtin/packages/libogg/package.py:
```diff
@@ -24,3 +30,43 @@ class Libogg(AutotoolsPackage):
+
+
+class AutotoolBuilder(AutotoolsBuilder):
+    pass
```
If you aren't overriding any functions, you should not have to instantiate the class here (same for CMakeBuilder)

johnwparent replied:
Ok, that was not clear. I'll remove.

scheibelp commented on var/spack/repos/builtin/packages/c-blosc/package.py:
```diff
@@ -43,6 +44,8 @@ class CBlosc(CMakePackage):
         return find_libraries("libblosc", root=self.prefix, recursive=True)
 
+
+class CMakeBuilder(CMakeBuilder):
```
It's fine to do this, but was it necessary? IMO it's ideal to avoid using the builder classes when there's only one possible choice.

johnwparent replied:
I was under the impression we wanted to convert all packages to using the builder class model. Is that not the case?

adamjstewart replied:
Only packages that support multiple build systems, not packages that support a single build system

johnwparent replied:
Ok, that's my misunderstanding then. Thankfully I did most of these ports atomically commit-wise so I'll just drop those commits. Thanks @adamjstewart

scheibelp commented on var/spack/repos/builtin/packages/double-conversion/package.py:
```diff
@@ -29,5 +30,7 @@ class DoubleConversion(CMakePackage):
     version("1.1.3", sha256="f0d1b8621592a3cf010c04c3e1c0f08455fc0fc7ee22e1583e2a63dc6d3e3871")
 
+
+class CMakeBuilder(CMakeBuilder):
```
(same question here as for `c-blosc`)

johnwparent replied:
All c-blosc comments can refer to that response and thread as the answer will likely be the same and I don't want to spam with the same comment over and over. If there is a unique case I will address it at that time.

johnwparent replied:
Issue addressed above.

scheibelp commented on var/spack/repos/builtin/packages/eigen/package.py:
```diff
@@ -68,16 +68,18 @@ class Eigen(CMakePackage):
+        return headers
+
+
+class CMakeBuilder(CMakeBuilder):
```
(same question here as for `c-blosc`)

scheibelp commented on var/spack/repos/builtin/packages/expat/package.py:
```diff
@@ -105,12 +106,11 @@ def url_for_version(self, version):
-    def configure_args(self):
-        spec = self.spec
+
+class CMakeBuilder(CMakeBuilder):
```
(same question here as for `c-blosc`)

scheibelp commented on var/spack/repos/builtin/packages/jsoncpp/package.py:
```diff
@@ -60,6 +61,8 @@ def patch(self):
         )
 
+
+class CMakeBuilder(CMakeBuilder):
```
(same question here as for `c-blosc`)

scheibelp commented on var/spack/repos/builtin/packages/libiconv/package.py:
```diff
@@ -28,7 +29,8 @@ class Libiconv(AutotoolsPackage, GNUMirrorPackage):
     patch("gets.patch", when="@1.14")
-    provides("iconv")
+    if sys.platform != "win32":
+        provides("iconv")
```
I think it would be clearer to add `conflicts(platform=windows)` since it looks like Spack shouldn't try to build this at all on Windows.

johnwparent replied:
Good idea. Will do.

scheibelp commented on var/spack/repos/builtin/packages/libpng/package.py:
```diff
@@ -3,10 +3,11 @@
 
 
-class Libpng(AutotoolsPackage):
+class Libpng(CMakePackage):
```
The PR description mentions

> libpng - to multibuild system

but it looks like this just switches libpng to use CMake on both Windows and Linux (which presumably would work fine).

johnwparent replied:
Must have gotten my own wire cross when creating the lists of which changes were made to what packages, it's in my multi build list but clearly that's not the case.

scheibelp commented on var/spack/repos/builtin/packages/libtheora/package.py:
```diff
@@ -33,14 +31,25 @@ class Libtheora(AutotoolsPackage):
+    depends_on("autoconf", type="build")
+    depends_on("automake", type="build")
+    depends_on("libtool", type="build")
+    depends_on("m4", type="build")
```
`depends_on` cannot be placed in the builder, it must appear in the package with something like `with when("build_system=autotools"):` (see `arpack-ng` and `libjpeg-turbo` for examples).

johnwparent replied:
Sounds good.

scheibelp commented on var/spack/repos/builtin/packages/libtheora/package.py:
```diff
@@ -3,10 +3,12 @@
 
 
-class Libtheora(AutotoolsPackage):
+class Libtheora(Package):
```
This should still inherit autotools. If this can build with scons in some cases (on Windows I assume) then it should also define a `build_system` variant. I thought `scons` worked on both Linux/Windows, so it would probably make sense to steer the default to be Autotools on linux and SCons on Windows.

scheibelp commented on var/spack/repos/builtin/packages/libtiff/package.py:
```diff
@@ -128,3 +128,15 @@ def configure_args(self):
 
         return args
+
+    def cmake_args(self):
```
Why does the AutotoolsBuilder have a cmake_args and cmake function defined?

scheibelp commented on var/spack/repos/builtin/packages/libxml2/package.py:
```diff
@@ -17,6 +21,8 @@ class Libxml2(AutotoolsPackage):
     url = "https://download.gnome.org/sources/libxml2/2.9/libxml2-2.9.13.tar.xz"
     list_url = "https://gitlab.gnome.org/GNOME/libxml2/-/releases"
 
+    phases = ["configure", "build", "install"]
```
You shouldn't have to define `phases` here: I think it's clearer to keep that in the builders.

scheibelp commented on var/spack/repos/builtin/packages/netcdf-c/package.py:
```diff
@@ -86,11 +88,11 @@ class NetcdfC(AutotoolsPackage):
-    depends_on("autoconf", type="build", when="@4.7.0,main")
-    depends_on("automake", type="build", when="@4.7.0,main")
-    depends_on("libtool", type="build", when="@4.7.0,main")
+    # depends_on("autoconf", type="build", when="@4.7.0,main")
```
This should be uncommented and likely inside of a `with when(build_system=autotools)` block.

scheibelp commented on var/spack/repos/builtin/packages/netcdf-cxx/package.py:
```diff
@@ -26,8 +27,10 @@ def libs(self):
-    def configure_args(self):
-        args = []
+
+class CMakeBuilder(CMakeBuilder):
```
The PR description mentions

> netcdf-cxx - to multibuild system

but it looks like this just replaces an Autotools build with a CMake-based build. If so, that's fine but a separate CMakeBuilder class is not necessary.

johnwparent replied:
Clearly my tracking of what was done where was severely disorganized. I'll fix the PR description.

scheibelp commented on var/spack/repos/builtin/packages/paraview/package.py:
```diff
@@ -341,6 +342,8 @@ def setup_run_environment(self, env):
                     env.prepend_path("PYTHONPATH", join_path(pv_pydir, "_vtk.zip"))
 
+
+class CMakeBuilder(CMakeBuilder):
```
(same question here as for `c-blosc`)

scheibelp commented on var/spack/repos/builtin/packages/pegtl/package.py:
```diff
@@ -32,6 +32,14 @@ class Pegtl(CMakePackage):
+            make("test", parallel=False)
+
+
+class CMakeBuilder(CMakeBuilder):
```
(same question here as for `c-blosc`)

scheibelp commented on var/spack/repos/builtin/packages/perl/package.py:
```diff
@@ -515,3 +347,173 @@ def test(self):
+    @run_after("install")
+    def symlink_windows(self):
+        if not is_windows:
+            return
```
I think it would be clearer to put this specific callback in the NmakeBuilder definition (since it only runs there).

johnwparent replied:
Ok, wasn't sure where callbacks could go.

scheibelp commented on var/spack/repos/builtin/packages/proj/package.py:
```diff
@@ -83,7 +83,7 @@ class Proj(AutotoolsPackage):
 
     # https://proj.org/install.html#build-requirements
-    depends_on("pkgconfig@0.9.0:", type="build", when="@6:")
+    # depends_on("pkgconfig@0.9.0:", type="build", when="@6:")
```
Remove commented-out dependency.

It looks like this now uses CMake everywhere (Linux and Windows) - that's fine if so just checking.

scheibelp commented on var/spack/repos/builtin/packages/protobuf/package.py:
```diff
@@ -95,6 +96,8 @@ class Protobuf(CMakePackage):
         sha256="fa1abf042eddc1b3b43875dc018c651c90cd1c0c5299975a818a1610bee54ab8",
     )
 
+    build_system("cmake", "autotools", default="cmake")
```
This used to use CMake on Linux. Is there a reason to add this? If this is needed, the package should inherit AutotoolsPackage as well. If not, then the separate CMakeBuilder is not needed (i.e. like for `c-blosc`).

scheibelp commented on var/spack/repos/builtin/packages/snappy/package.py:
```diff
@@ -23,6 +24,8 @@ class Snappy(CMakePackage):
     patch("link_gtest.patch", when="@:1.1.8")
 
+
+class CMakeBuilder(CMakeBuilder):
```
(same question as for `c-blosc`)

scheibelp commented on var/spack/repos/builtin/packages/vtk/package.py:
```diff
@@ -178,6 +179,8 @@ def setup_build_environment(self, env):
                 env.append_flags("CXXFLAGS", "-DH5_USE_18_API")
 
+
+class CMakeBuilder(CMakeBuilder):
```
(same question as for `c-blosc`)

scheibelp commented on var/spack/repos/builtin/packages/zfp/package.py:
```diff
@@ -121,7 +122,8 @@ class Zfp(CMakePackage, CudaPackage):
 
-    # CMake options
+
+class CMakeBuilder(CMakeBuilder):
```
(same question as `c-blosc`)

scheibelp commented on var/spack/repos/builtin/packages/zstd/package.py:
```diff
@@ -55,36 +58,20 @@ class Zstd(MakefilePackage):
 
-        # install the library
-        make(*lib_args)
+class CMakeBuilder(CMakeBuilder):
```
If this now uses CMake on all platforms, a separate CMakeBuilder is not needed (like `c-blosc`).

scheibelp commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -1206,3 +903,350 @@ def test(self):
+
+            # Ensure that dbm module works
+            if "+dbm" in spec:
+                self.command("-c", "import dbm")
```
I don't see the Windows build handling `~dbm`. If it's not supported or we don't want to handle that for now, then I think `conflicts(~dbm, when=platform=windows)` would be appropriate.

johnwparent replied:
It never has, this PR does nothing to change that behavior. The python build script on Windows has an incredibly limited CLI so many of the variants will have no impact one way or another on Windows. I added the conflict just to represent the fact that whether or not you specify the variant, you're getting dbm on Windows.

scheibelp reviewed (changes requested):
Most of these comments are very minor: I think when a package doesn't define multiple build systems it doesn't need to add a BuildSystem definition (I refer to `c-blosc` often as an example). Let me know if I overlooked some reason that needs to happen.

scheibelp commented on var/spack/repos/builtin/packages/zlib/package.py:
```diff
@@ -138,4 +138,4 @@ def installtree(dst, tree):
                     install(file, install_dst)
 
-        installtree(self.prefix, install_tree)
+        installtree(prefix, install_tree)
```
We have a function called `install_tree` but not one called `installtree` - has this been confirmed to work?

johnwparent replied:
It's installed on my local system from this pr.... But you're absolutely correct, that's not the name of the function defined in the filesystem module. Very confused about that, will have to investigate...

johnwparent replied:
The `installtree` method was something I implemented some time ago when porting zlib for the first time. I'm not sure if its still needed, or why it was implemented in the first place as the individual commit message was lost to the Windows super PR. I'll try replacing it with the `filesystem.install_tree` method and see if that works.

scheibelp commented on var/spack/repos/builtin/packages/proj/package.py:
```diff
@@ -83,30 +83,16 @@ class Proj(AutotoolsPackage):
+        return [
+            self.define("PROJ_LIB", join_path(self.stage.source_path, "nad")),
+            self.define_from_variant("ENABLE_TIFF", "tiff"),
+        ]
```
Are there similar options for enabling `curl`? If not then it might make sense to remove the `+curl` variant (unless CMake detects this automatically)

johnwparent replied:
I didn't see them. I think I'll refactor this one to retain the autotools build on non Windows platforms pending further investigation at a later date.

scheibelp commented on lib/spack/spack/build_systems/msbuild.py:
```diff
@@ -0,0 +1,123 @@
+                *self.msbuild_install_args(), self.define_targets(*self.install_targets)
+            )
+
+    # def setup_dependent_package(self, m, dep_spec):
```
This seems preferable to setting the variables in `build_environment`: i.e. even if we don't have an `msbuild`/`cscript` package, recording that this is the build system that needs them is worthwhile. Do any problems occur when you uncomment this?

(likewise for the section in `nmake.py`)

scheibelp replied:
If build systems other than `msbuild` use these, then I can see the appeal of defining them in `build_environment` (similar to why `make` is defined there).

johnwparent replied:
Only msbuild/nmake uses these, although there are currently generic classes using them, but those should be refactored. However, when I switched to using it this way, all nmake and msbuild `build`  builder phases fail because `nmake` or `msbuild` respectively are not defined, and it seems like this method is not invoked before the build stage. It seems as though other build tools either have packages to represent them, allowing for these tools to be defined when the package is loaded, or are defined in the `build_environment` module. According to the docs, `setup_dependent_packages` is only invoked before `install`, which makes it unsuitable for use in this case, as we need it during the `build` phase. If this is not the case, then something is broken and the docs need updating. If this is the case, can we implement something that does run before `build` or `configure`? Or maybe this is more configurable than I think and I'm not seeing it in the build system docs.

scheibelp replied:
`setup_dependent_packages` (and all `*setup*` functions) should run before all phases and it is indeed a bug if they are not run before the `build` phase - can you mention the error message here?

johnwparent replied:
The error:
```
Traceback (most recent call last):
  File "C:\spack_win\spack\lib\spack\spack\build_environment.py", line 1053, in _setup_pkg_and_run
    return_value = function(pkg, kwargs)
  File "C:\spack_win\spack\lib\spack\spack\installer.py", line 2081, in build_process
    return installer.run()
  File "C:\spack_win\spack\lib\spack\spack\installer.py", line 1945, in run
    self._real_install()
  File "C:\spack_win\spack\lib\spack\spack\installer.py", line 2041, in _real_install
    phase_fn.execute()
  File "C:\spack_win\spack\lib\spack\spack\builder.py", line 420, in execute
    self.phase_fn(pkg, pkg.spec, pkg.prefix)
  File "C:\spack_win\spack\var\spack\repos\builtin\packages\winlibiconv\package.py", line 37, in build
NameError: name 'nmake' is not defined
```

When I include a line that, if the `setup_dependent_package` is invoked will exit with failure, nothing happens and execution continues as normal. Perhaps I'm doing something wrong because even if I remove the contents of the build method, the install method fails the same way, so either I've done something wrong, or this method is never being invoked. Considering I have a `raise` and an `exit(1)` in that stanza and nothing happens, I feel like it might never be invoked.

scheibelp replied:
Whoops sorry it just occurred to me: this should go into `setup_package`, not `setup_dependent_package`: if you have A->B, then setup_dependent_package is run for B when building A, but not for B when building B. Does that make sense?

johnwparent replied:
It does make sense, and I did think it was a bit of a bizarre phrasing for the function name. I just assumed it was for "packages dependent on the builder". That being said, switching to `setup_package` changes nothing, nor can I find a reference to that method being invoked anywhere except for a `build_environment` context. It's a method in `build_environment` that invokes some environmental information about the package, but doesn't seem to be invoked by anything that would call our packages `setup_package` method, at least according to a grep of the codebase. 

From what I can tell, we rely on `setup_dependent_package` for almost every build system but autotools/makefiles and maven, as they all have representative packages. Autotools pulls from `build_environment` as we know, and maven runs a `which` during the install phase. I'm happy to add stub packages for nmake and msbuild that don't install, but do allow for external detection but being required to do that feels like a bug with the way things are currently structured.

Maybe we should sync about this later today?

scheibelp replied:
@alalazo it turns out that for now we need to keep defining `cscript`/`msbuild` in `build_environment`:

* We're able to customize CMake because packages which use it depend on the `cmake` Spack package, which does `setup_dependent_package`
* There is currently no Spack package which provides `msbuild`/`cscript`
* We don't have a way for build systems to indicate that they should modify the package module, i.e. there is no `setup_package` analog to `setup_dependent_package` (to avoid confusion, there *is* a method called `setup_package` but it is not used for this purpose)

For now I think the path forward is to:

* Define `msbuild`/`cscript` in `build_environment`
* Add a TODO message there
* In a follow-up PR, define a dependency package, or update the API to allow build systems to modify the module

Does that sound reasonable?

scheibelp commented on var/spack/repos/builtin/packages/adios2/package.py:
```diff
@@ -161,6 +172,45 @@ def setup_build_environment(self, env):
             env.set("FFLAGS", "-Ccpp")
 
+    @property
+    def libs(self):
```
It appears that this and `setup_environment` moved but were not modified at all - can you move them back to minimize the diff (also helps with `git blame`)?

scheibelp commented on var/spack/repos/builtin/packages/expat/package.py:
```diff
@@ -112,12 +112,19 @@ def url_for_version(self, version):
+    #     ]
+    #     if "+libbsd" in spec and "@2.2.1:" in spec:
+    #         args.append("--with-libbsd")
+    #     return args
```
Commented code should be removed. In this case it looks like not all the `configure` options have been ported to `cmake`. I think it might be worth splitting this package to another PR to resolve this issue.

johnwparent replied:
Yes, I left configure sections that I was not able to replicate 1-1 in cmake's configure stage commented out for the most part to provide context for the missing args. Will amend as discussed on Slack.

scheibelp commented on var/spack/repos/builtin/packages/freetype/package.py:
```diff
@@ -47,14 +47,7 @@ def headers(self):
+
+class CMakeBuilder(CMakeBuilder):
+    def cmake_args(self):
+        return [self.define("FT_WITH_ZLIB", True), self.define("FT_WITH_BZIP2", True)]
```
Is there a similar CMake option for `libpng`?

johnwparent replied:
I didn't see one bizarrely enough, I thought that was weird as well. I'll fix as we discussed over slack.

scheibelp commented on var/spack/repos/builtin/packages/libpng/package.py:
```diff
@@ -26,17 +26,8 @@ class Libpng(AutotoolsPackage):
-    def check(self):
-        # Libpng has both 'check' and 'test' targets that are aliases.
-        # Only need to run the tests once.
-        make("check")
```
Does CMake have a similar option for running tests/checks after the build?

johnwparent replied:
It depends if the requisite target's have been defined/ctest was used. For this package in particular I don't know. I'll be addressing as discussed.

scheibelp commented on var/spack/repos/builtin/packages/libtheora/package.py:
```diff
@@ -17,13 +21,16 @@ class Libtheora(AutotoolsPackage):
+        depends_on("libtool", type="build")
+        depends_on("m4", type="build")
+
+    build_system("scons", "autotools", default="autotools" if sys.platform == "win32" else "scons")
```
I think you mean `default="scons" if sys.platform == "win32" else "autotools"`

johnwparent replied:
Yes, I reversed that, good catch,

scheibelp commented on var/spack/repos/builtin/packages/lz4/package.py:
```diff
@@ -44,6 +49,39 @@ def url_for_version(self, version):
+            filter_file("-pedantic", "", "Makefile")
+
+    @run_after("install")
+    def darwin_fix(self):
```
I think this will need to move to a class that is inherited by the CMakeBuild and Makefile builder (like `libxml2`).

scheibelp commented on var/spack/repos/builtin/packages/netcdf-c/package.py:
```diff
@@ -142,12 +145,62 @@ class NetcdfC(AutotoolsPackage):
 
     filter_compiler_wrappers("nc-config", relative_root="bin")
 
+    build_system("cmake", "autotools", default="cmake")
```
To minimize the effect of this PR on non-windows, IMO it would be worthwhile to have default build system match the prior chosen one where possible.

scheibelp commented on var/spack/repos/builtin/packages/netcdf-c/package.py:
```diff
@@ -142,12 +145,62 @@ class NetcdfC(AutotoolsPackage):
+        # Some packages, e.g. ncview, refuse to build if the compiler path returned by nc-config
+        # differs from the path to the compiler that the package should be built with. Therefore,
+        # we have to shadow nc-config from self.prefix.bin, which references the real compiler,
+        # with a backed up version, which references Spack compiler wrapper.
```
This looks like it would apply for the CMake builder as well

scheibelp commented on var/spack/repos/builtin/packages/netcdf-cxx/package.py:
```diff
@@ -26,8 +26,8 @@ def libs(self):
+        args = [self.define_from_variant("USE_NETCDF4", "netcdf4")]
         if "+netcdf4" in self.spec:
             # There is no clear way to set this via configure, so set the flag
             # explicitly
```
Does CMake also recognize `CPPFLAGS=` as an option?

scheibelp commented on var/spack/repos/builtin/packages/zstd/package.py:
```diff
@@ -55,36 +57,18 @@ class Zstd(MakefilePackage):
-
-        # Tested %nvhpc@22.3. No support for -MP
-        if "%nvhpc" in self.spec:
-            args.append("DEPFLAGS=-MT $@ -MMD -MF")
```
It looks like this didn't get ported over to the CMake version.

johnwparent replied:
Discussed over slack. Will amend as we decided there.

scheibelp commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -1219,3 +800,464 @@ def test(self):
+                copy(lib, prefix.DLLs)
+        static_libraries = glob.glob("%s\\*.lib")
+        for lib in static_libraries:
+            copy(lib, prefix.libs)
```
Am I correct that this PR makes no changes in this package other than to move functionality into respective builder classes?

johnwparent replied:
In general, yes. I was under the impression that we wanted to move packages like this to multiple builders, and as its a dep of paraview, I thought it was warranted under the scope of this PR.

scheibelp commented on var/spack/repos/builtin/packages/perl/package.py:
```diff
@@ -457,61 +367,170 @@ def filter_config_dot_pm(self):
+            lnk_path = os.path.join(self.prefix.bin, f)
+            src_path = os.path.join(win_install_path, f)
+            if not os.path.exists(lnk_path):
+                symlink(src_path, lnk_path)
```
Similar question here as for Python: does this PR make no other changes to this package other than to move functionality into respective builders?

johnwparent replied:
There are also changes w.r.t. to how Perl gets its MSVC toolchain version as I refactored the underlying OS class. But for the most part correct. I have a similar response to this as I do to the python comment.

scheibelp reviewed (changes requested):
Where possible, I think this PR might benefit from a smoother transition for build systems on non-Windows by preserving the old build system where possible. This would resolve some of the comments (i.e. in cases where I ask whether a given `configure` option has an anolog in CMake); beyond that though it might mitigate unforeseen issues with changing the build system.

johnwparent committed:
Add CMake builder to zstd

johnwparent committed:
NASM fixup

johnwparent committed:
hdf5 general improvements

johnwparent committed:
Proper dep for Paraview on libxml2

johnwparent committed:
Proper scons executable on Windows

johnwparent committed:
Finalize xz

xz - mulit build system

style xz

update xz

update xz

Refactor xz build

Proper xz toolchain override

dereference builder

Fixup xz

johnwparent committed:
libtheora

johnwparent committed:
Further libxml2 support

fixup libxml

better libxml2

Modernize libxml2

libxml2 - multi build system

update libxml2

update libxml2

libxml2 proper pkgconf dep

libxml2 correct nmake import

johnwparent committed:
Proper msmpi detection stanza

johnwparent committed:
update wgl

Add default gl provider

Prevent apple-gl on Windows platform

Improve WGL headers/libs detection

johnwparent committed:
freetype - multi build system

free type style

johnwparent committed:
update proj

johnwparent committed:
update zlib

johnwparent committed:
update netcdf-c

NetCDF-C patches

Fixup netcdf HDF5+zlib detection

Netcdf's cmake is unable to detect that HDF5 was compiled with zlib
Add a patch that allows for said detection

netcdf-c patch ill formed cmakelists

Add netcdf linker patch

johnwparent committed:
netcdf-c

johnwparent committed:
netcdf-cxx port

johnwparent committed:
Rename OpenGL Binary on Win

johnwparent committed:
NMake builder: Proper makefile root invocation

johnwparent committed:
Fixup libtheora port

johnwparent committed:
Improvements to msvc

johnwparent committed:
Update libogg port

johnwparent committed:
xz: proper pkg dereference

johnwparent committed:
msmpi: proper compiler pkg dereference

johnwparent committed:
Fixup libtheora build structure modeling

johnwparent committed:
Fixup xz install phase

johnwparent committed:
Add mpi compiler wrappers to msmpi

johnwparent commented:
Closed in favor of multiple smaller PRs
